src/System.Security.Principal.Windows/tests/WindowsIdentityTests.Impersonate.cs

```diff
@@ -0,0 +1,709 @@
+                    Assert.Equal(_fixture.TestAccount2.AccountName, InteropHelper.GetCurrentUserName(), ignoreCase: true);
+
+                    await Task.Run(async () =>
+                    {
+                        Assert.Equal(_fixture.TestAccount2.AccountName, InteropHelper.GetCurrentUserName(), ignoreCase: true);
+
+                        using (WindowsImpersonationContext windowsImpersonationContext2 = WindowsIdentity.Impersonate(_fixture.TestAccount1.AccountTokenHandle.DangerousGetHandle()))
+                        {
+                            Assert.Equal(_fixture.TestAccount1.AccountName, InteropHelper.GetCurrentUserName(), ignoreCase: true);
+
+                            await Task.Run(() =>
+                            {
+                                WindowsIdentity.RunImpersonated(_fixture.TestAccount2.AccountTokenHandle, () =>
+                                {
+                                    Assert.Equal(_fixture.TestAccount2.AccountName, InteropHelper.GetCurrentUserName(), ignoreCase: true);
+                                });
+                            });
+
+                            Assert.Equal(_fixture.TestAccount1.AccountName, InteropHelper.GetCurrentUserName(), ignoreCase: true);
+                        }
+
+                        Assert.Equal(_fixture.TestAccount2.AccountName, InteropHelper.GetCurrentUserName(), ignoreCase: true);
+                    });
+
+                    Assert.Equal(_fixture.TestAccount2.AccountName, InteropHelper.GetCurrentUserName(), ignoreCase: true);
+                });
+
+                Assert.Equal(_fixture.TestAccount1.AccountName, InteropHelper.GetCurrentUserName(), ignoreCase: true);
+            });
+
+            Assert.Equal(_fixture.TestAccount1.AccountName, InteropHelper.GetCurrentUserName(), ignoreCase: true);
+
+        }
+    }
+}
+
+// For context flow tests we use custom task scheduler to be sure that 
+// we're using "clean identity token thread". 
+// If we do bug it's possible return impersonated thread to thread pool
+// and pick up the same for continuation task generating false positive.
+public class CustomTaskScheduler : TaskScheduler
+{
+    public ConcurrentBag<(string, int)> ThreadsData { get; } = new ConcurrentBag<(string, int)>();
+
+    private volatile int _outstandingThread = 0;
+
+    protected override IEnumerable<Task> GetScheduledTasks()
+    {
+        throw new NotImplementedException();
+    }
+
+    protected override bool TryDequeue(Task task)
+    {
+        throw new NotImplementedException();
+    }
+
+    protected override void QueueTask(Task task)
+    {
+        Interlocked.Increment(ref _outstandingThread);
+        var thread = new Thread(_ =>
+        {
+            TryExecuteTask(task);
+
+            // we verify if thread is "clean"
+            ThreadsData.Add((InteropHelper.GetCurrentUserName(), Thread.CurrentThread.ManagedThreadId));
+
+            Interlocked.Decrement(ref _outstandingThread);
+        });
```

Comment:
These should have { IsBackground = true }.

Reply:
ok

Reply:
done